system/mp/mpcomm_mpi.cpp

```diff
@@ -0,0 +1,302 @@
+                hpcc_mpi::releaseComm(req);
+            }
+        }
+       
+        return true;
+    }
+
+    bool recv(CMessageBuffer &mbuf, rank_t srcrank, mptag_t tag, rank_t *sender, unsigned timeout=MP_WAIT_FOREVER){
+        CTimeMon tm(timeout);
+        hpcc_mpi::CommRequest req = hpcc_mpi::readData(srcrank, tag, mbuf, *group, true);
+        unsigned remaining;
+        hpcc_mpi::CommStatus stat;
+        do {
+            usleep(100);
+            stat = hpcc_mpi::getCommStatus(req);
+        } while (stat == hpcc_mpi::CommStatus::INCOMPLETE && !tm.timedout(&remaining));
+//        printf("recv while existed %d\n",stat);
+        hpcc_mpi::releaseComm(req);
+//        printf("req released %d\n",stat);
+        return (stat == hpcc_mpi::CommStatus::SUCCESS);
+    }
+    
+    void barrier(void){
+        hpcc_mpi::barrier(*group);
+    }
+
+    unsigned probe(rank_t srcrank, mptag_t tag, rank_t *sender, unsigned timeout=0){
+        if (hpcc_mpi::hasIncomingMessage(srcrank, tag, *group)){
+            if (sender)
+                *sender = srcrank;
+        }
+        return false;
+    }
+    
+    void flush(mptag_t tag){
+        // Handled by MPI
+    }
+
+    bool sendRecv(CMessageBuffer &mbuff, rank_t sendrank, mptag_t sendtag, unsigned timeout=MP_WAIT_FOREVER){
+        //TODO share timeout between send/recv?
+        mptag_t replytag = createReplyTag();
+        CTimeMon tm(timeout);
+        mbuff.setReplyTag(replytag);
+        unsigned remaining;
+        if (tm.timedout(&remaining))
+            return false;
+        if (!send(mbuff,sendrank,sendtag,remaining)||tm.timedout(&remaining))
+            return false;
+        mbuff.clear();
+        return recv(mbuff,sendrank,replytag,NULL,remaining);
+    }
+
+    bool reply(CMessageBuffer &mbuf, unsigned timeout=MP_WAIT_FOREVER){
+        mptag_t replytag = mbuf.getReplyTag();
+        rank_t dstrank = group->rank(mbuf.getSender());
+        if (dstrank!=RANK_NULL) {
+            if (send (mbuf, dstrank, replytag,timeout)) {
+                mbuf.setReplyTag(TAG_NULL);
+                return true;
+            }
+            return false;
+        }
+        return false;
+    }
+
+    void cancel(rank_t srcrank, mptag_t tag){
+        assertex(srcrank!=RANK_NULL);       
+        hpcc_mpi::CommRequest req = group->getCommRequest(srcrank, tag);
+        if (req >= 0){
+            hpcc_mpi::cancelComm(req);
+        }
+    }
+
+    bool verifyConnection(rank_t rank,  unsigned timeout){
+        UNIMPLEMENTED;
+    }
+
+    bool verifyAll(bool duplex, unsigned timeout){
+        UNIMPLEMENTED;
+    }
+    
+    void disconnect(INode *node){
+        UNIMPLEMENTED;
+    }
+
+    IGroup &queryGroup() { 
+        UNIMPLEMENTED;
+    }
+    
+    IGroup *getGroup()  { 
+        return group;
+    }
+
+    NodeCommunicator(IGroup *_group){
+        //TODO throw meaningful exception if _group is not NodeGroup type
+        this->group = (NodeGroup*) _group;
```

Comment:
MPI implementation of ICommunicator interface require MPI aware IGroup implementation. Hence why the casting. I was trying to think of a better way to solve this. Couldn't this of one. Perhaps I'm lacking my understanding here on how this is done properly. But I agree that this definitely require some attention.
Someone from outside trying to use the MP project first creates an IGroup object using "createIGroup(...)" function and calls "createCommunicator(IGroup *group, ...)" function to create the communicator. Thus by the time the communicator gets created it has no choice but to expect an IGroup type even though we have the control over how "createCommunicator(...)" is implemented. And I didn't want to complicate things by introducing new interfaces like IMPIGroup this early in the project since we didn't know what restrictions we'd face later on.

Reply:
>MPI implementation of ICommunicator interface require MPI aware IGroup implementation. 

it does at the moment, and there may still be key differences, but I think they'll become more obvious what is really required when the comms functionality moves in NodeCommunicator and out of the IGroup implementation.